File: schemakeeper-avro/src/main/scala/schemakeeper/avro/AvroSchemaUtils.scala
package schemakeeper.avro


import org.apache.avro.Schema
import org.apache.avro.generic.GenericContainer

object AvroSchemaUtils {
  private val primitiveTypes = Map[Schema.Type, Schema](
    Schema.Type.STRING -> Schema.create(Schema.Type.STRING),
    Schema.Type.BYTES -> Schema.create(Schema.Type.BYTES),
    Schema.Type.INT -> Schema.create(Schema.Type.INT),
    Schema.Type.LONG -> Schema.create(Schema.Type.LONG),
    Schema.Type.FLOAT -> Schema.create(Schema.Type.FLOAT),
    Schema.Type.DOUBLE -> Schema.create(Schema.Type.DOUBLE),
    Schema.Type.BOOLEAN -> Schema.create(Schema.Type.BOOLEAN),
    Schema.Type.NULL -> Schema.create(Schema.Type.NULL)
  )

  def getSchema(value: Any): Schema = value match {
    case null => primitiveTypes(Schema.Type.NULL)
    case _: String => primitiveTypes(Schema.Type.STRING)
    case _: Int => primitiveTypes(Schema.Type.INT)
    case _: Short => primitiveTypes(Schema.Type.INT)
    case _: Byte => primitiveTypes(Schema.Type.INT)
    case _: Long => primitiveTypes(Schema.Type.LONG)
    case _: Float => primitiveTypes(Schema.Type.FLOAT)
    case _: Double => primitiveTypes(Schema.Type.DOUBLE)
    case _: Boolean => primitiveTypes(Schema.Type.BOOLEAN)
    case _: Array[Byte] => primitiveTypes(Schema.Type.BYTES)
    case c: GenericContainer => c.getSchema
    case _ => throw new IllegalArgumentException("Unsupported avro type")
  }

  def isPrimitive(value: Any): Boolean = value match {
    case null => true
    case _: String => true
    case _: Int => true
    case _: Short => true
    case _: Byte => true
    case _: Long => true
    case _: Float => true
    case _: Double => true
    case _: Boolean => true
    case _: Array[Byte] => true
    case _: GenericContainer => false
    case _ => throw new IllegalArgumentException("Unsupported avro type")
  }

  def isPrimitive(schema: Schema): Boolean = primitiveTypes.contains(schema.getType)
}

File: schemakeeper-common/src/main/java/schemakeeper/schema/AvroSchemaUtils.java
package schemakeeper.schema;

import org.apache.avro.Schema;
import org.apache.avro.generic.GenericRecord;

import java.util.HashMap;
import java.util.Map;

public class AvroSchemaUtils {
    private static final Map<Schema.Type, Schema> PRIMITIVE_TYPES;

    static {
        PRIMITIVE_TYPES = new HashMap<>();

        PRIMITIVE_TYPES.put(Schema.Type.STRING, Schema.create(Schema.Type.STRING));
        PRIMITIVE_TYPES.put(Schema.Type.BYTES, Schema.create(Schema.Type.BYTES));
        PRIMITIVE_TYPES.put(Schema.Type.INT, Schema.create(Schema.Type.INT));
        PRIMITIVE_TYPES.put(Schema.Type.LONG, Schema.create(Schema.Type.LONG));
        PRIMITIVE_TYPES.put(Schema.Type.FLOAT, Schema.create(Schema.Type.FLOAT));
        PRIMITIVE_TYPES.put(Schema.Type.DOUBLE, Schema.create(Schema.Type.DOUBLE));
        PRIMITIVE_TYPES.put(Schema.Type.BOOLEAN, Schema.create(Schema.Type.BOOLEAN));
        PRIMITIVE_TYPES.put(Schema.Type.NULL, Schema.create(Schema.Type.NULL));
    }

    public static Schema getSchema(Object value) {
        if (value == null) {
            return PRIMITIVE_TYPES.get(Schema.Type.NULL);
        } else if (value instanceof String) {
            return PRIMITIVE_TYPES.get(Schema.Type.STRING);
        } else if (value instanceof Integer) {
            return PRIMITIVE_TYPES.get(Schema.Type.INT);
        } else if (value instanceof Short) {
            return PRIMITIVE_TYPES.get(Schema.Type.INT);
        } else if (value instanceof Byte) {
            return PRIMITIVE_TYPES.get(Schema.Type.INT);
        } else if (value instanceof Long) {
            return PRIMITIVE_TYPES.get(Schema.Type.LONG);
        } else if (value instanceof Float) {
            return PRIMITIVE_TYPES.get(Schema.Type.FLOAT);
        } else if (value instanceof Double) {
            return PRIMITIVE_TYPES.get(Schema.Type.DOUBLE);
        } else if (value instanceof Boolean) {
            return PRIMITIVE_TYPES.get(Schema.Type.BOOLEAN);
        } else if (value instanceof byte[]) {
            return PRIMITIVE_TYPES.get(Schema.Type.BYTES);
        } else if (value instanceof GenericRecord) {
            return ((GenericRecord) value).getSchema();
        } else {
            throw new IllegalArgumentException("Unsupported avro type");
        }
    }

    public static boolean isPrimitive(Object value) {
        if (value == null) {
            return true;
        } else if (value instanceof String) {
            return true;
        } else if (value instanceof Integer) {
            return true;
        } else if (value instanceof Short) {
            return true;
        } else if (value instanceof Byte) {
            return true;
        } else if (value instanceof Long) {
            return true;
        } else if (value instanceof Float) {
            return true;
        } else if (value instanceof Double) {
            return true;
        } else if (value instanceof Boolean) {
            return true;
        } else if (value instanceof byte[]) {
            return true;
        } else if (value instanceof GenericRecord) {
            return false;
        } else {
            throw new IllegalArgumentException("Unsupported avro type");
        }
    }

    public static boolean isPrimitive(Schema schema) {
        return PRIMITIVE_TYPES.containsKey(schema.getType());
    }

    public static Schema parseSchema(String schema) {
        Schema.Parser parser = new Schema.Parser();
        return parser.parse(schema);
    }
}